src/caffe/layers/center_loss_layer.cpp

```diff
@@ -57,54 +59,64 @@ void CenterLossLayer<Dtype>::Forward_cpu(const vector<Blob<Dtype>*>& bottom,
   const Dtype* label = bottom[1]->cpu_data();
   const Dtype* center = this->blobs_[0]->cpu_data();
   Dtype* distance_data = distance_.mutable_cpu_data();
-  
   // the i-th distance_data
   for (int i = 0; i < M_; i++) {
     const int label_value = static_cast<int>(label[i]);
     // D(i,:) = X(i,:) - C(y(i),:)
-    caffe_sub(K_, bottom_data + i * K_, center + label_value * K_, distance_data + i * K_);
+    caffe_sub(K_, bottom_data + i * K_,
+              center + label_value * K_, distance_data + i * K_);
   }
-  Dtype dot = caffe_cpu_dot(M_ * K_, distance_.cpu_data(), distance_.cpu_data());
+  Dtype dot = caffe_cpu_dot(M_ * K_, distance_.cpu_data(),
+                            distance_.cpu_data());
   Dtype loss = dot / M_ / Dtype(2);
   top[0]->mutable_cpu_data()[0] = loss;
 }
 
 template <typename Dtype>
 void CenterLossLayer<Dtype>::Backward_cpu(const vector<Blob<Dtype>*>& top,
-    const vector<bool>& propagate_down,
-    const vector<Blob<Dtype>*>& bottom) {
+                                          const vector<bool>& propagate_down,
+                                          const vector<Blob<Dtype>*>& bottom) {
   // Gradient with respect to centers
   if (this->param_propagate_down_[0]) {
     const Dtype* label = bottom[1]->cpu_data();
     Dtype* center_diff = this->blobs_[0]->mutable_cpu_diff();
     Dtype* variation_sum_data = variation_sum_.mutable_cpu_data();
+    int* count_data = count_.mutable_cpu_data();
+
     const Dtype* distance_data = distance_.cpu_data();
 
     // \sum_{y_i==j}
     caffe_set(N_ * K_, (Dtype)0., variation_sum_.mutable_cpu_data());
-    for (int n = 0; n < N_; n++) {
-      int count = 0;
-      for (int m = 0; m < M_; m++) {
-        const int label_value = static_cast<int>(label[m]);
-        if (label_value == n) {
-          count++;
-          caffe_sub(K_, variation_sum_data + n * K_, distance_data + m * K_, variation_sum_data + n * K_);
-        }
-      }
-      caffe_axpy(K_, (Dtype)1./(count + (Dtype)1.), variation_sum_data + n * K_, center_diff + n * K_);
+    caffe_set(N_, 0 , count_.mutable_cpu_data());
+    caffe_set(N_ * K_, (Dtype)0., this->blobs_[0]->mutable_cpu_diff());
```

Comment:
Blob's diff should not be reseted in Backward().
It will hurt the gradient accumulation when iter_size > 1. See discussion in https://groups.google.com/forum/#!searchin/caffe-users/iter_size|sort:relevance/caffe-users/PMbycfbpKcY/FTBiMKunEQAJ